src/simularium/FrameRecorder.ts

```diff
@@ -0,0 +1,144 @@
+    }
+
+    private async setup(): Promise<void> {
+        const canvas = this.getCanvas();
+        if (canvas) {
+            try {
+                // VideoEncoder sends chunks of frame data to the muxer.
+                // Previously made one encoder in the constructor but
+                // making a new one during setup() prevents a bug where
+                // frames returned blank (stale canvas reference?)
+                this.encoder = new VideoEncoder({
+                    output: (chunk, meta) => {
+                        if (this.isRecording && this.muxer) {
+                            this.muxer?.addVideoChunk(chunk, meta);
+                        }
+                    },
+                    error: (error) => {
+                        console.error("Encoder error:", error);
+                    },
+                });
+                const config: VideoEncoderConfig = {
+                    codec: "avc1.420028",
+                    width: canvas.width,
+                    height: canvas.height,
+                };
+                const { supported, config: supportedConfig } =
+                    await VideoEncoder.isConfigSupported(config);
+                if (supported && supportedConfig) {
+                    this.encoder.configure(config);
+                } else {
+                    throw new Error("Unsupported video encoder configuration");
+                }
+                // Muxer will handle the conversion from raw video data to mp4
+                this.muxer = new Muxer({
+                    target: new ArrayBufferTarget(),
+                    video: {
+                        codec: "avc",
+                        width: canvas.width,
+                        height: canvas.height,
+                    },
+                });
+                this.frameIndex = 0;
+            } catch (error) {
+                throw new Error("Error setting up video encoder: " + error);
+            }
+        }
+    }
+
+    public async start(): Promise<void> {
+        try {
+            await this.setup();
+            this.isRecording = true;
+        } catch (e) {
+            console.log("setup failed", e);
+            return;
+        }
+    }
+
+    public async stop(): Promise<void> {
+        this.isRecording = false;
+        await this.onCompletedRecording();
+    }
+
+    public onFrame(): void {
+        if (!this.isRecording) {
+            return;
+        }
+        if (this.encoder) {
+            if (this.encoder.encodeQueueSize > 2) {
+                console.log("Dropping frame, too many frames in flight");
+                return;
+            }
+            const canvas = this.getCanvas();
+            if (canvas) {
+                // Add a keyframe every second: https://en.wikipedia.org/wiki/Key_frame
+                const keyFrame = this.frameIndex % this.frameRate === 0;
+                const timestampMicroseconds =
+                    (this.frameIndex / this.frameRate) * 1_000_000;
```

Comment:
Nit: I'm not sure I get the math here.  Because the calling code is already controlling the frame rate, could this be something more like Date.now() - startTime ?  I don't know that anything needs to change because maybe what I'm suggesting is really equivalent.  Are the timestamps supposed to be perfectly regular or can they be the true wall clock time?

Reply:
The `encoder` requires that the time stamps start at 0 and go up from there. If it weren't for that then you could probably use `Date.now() * 1000 - startTime`